crates/matrix-sdk/src/event_cache/room/mod.rs

```diff
@@ -749,182 +806,389 @@ mod private {
-                    current = next;
+                EventCacheStoreLockState::Dirty(store_guard) => {
+                    let mut guard = RoomEventCacheStateLockWriteGuard {
+                        state: state_guard,
+                        store: store_guard,
+                    };
+
+                    // Force to reload by shrinking to the last chunk.
+                    let updates_as_vector_diffs = guard.force_shrink_to_last_chunk().await?;
+
+                    // All good now, mark the cross-process lock as non-dirty.
+                    EventCacheStoreLockGuard::clear_dirty(&guard.store);
+
+                    // Now let the world know about the reload.
+                    if !updates_as_vector_diffs.is_empty() {
+                        // Notify observers about the update.
+                        let _ = guard.state.update_sender.send(
+                            RoomEventCacheUpdate::UpdateTimelineEvents {
+                                diffs: updates_as_vector_diffs,
+                                origin: EventsOrigin::Cache,
+                            },
+                        );
+
+                        // Notify observers about the generic update.
+                        let _ =
+                            guard.state.generic_update_sender.send(RoomEventCacheGenericUpdate {
+                                room_id: guard.state.room_id.clone(),
+                            });
+                    }
+
+                    Ok(guard)
                 }
             }
+        }
+    }
+
+    /// The read lock guard returned by [`RoomEventCacheStateLock::read`].
+    pub struct RoomEventCacheStateLockReadGuard<'a> {
+        /// The per-thread read lock guard over the
+        /// [`RoomEventCacheStateLockInner`].
+        state: RwLockReadGuard<'a, RoomEventCacheStateLockInner>,
+
+        /// The cross-process lock guard over the store.
+        store: EventCacheStoreLockGuard,
+    }
+
+    /// The write lock guard return by [`RoomEventCacheStateLock::write`].
+    pub struct RoomEventCacheStateLockWriteGuard<'a> {
+        /// The per-thread write lock guard over the
+        /// [`RoomEventCacheStateLockInner`].
+        state: RwLockWriteGuard<'a, RoomEventCacheStateLockInner>,
+
+        /// The cross-process lock guard over the store.
+        store: EventCacheStoreLockGuard,
+    }
+
+    impl<'a> RoomEventCacheStateLockWriteGuard<'a> {
+        /// Synchronously downgrades a write lock into a read lock.
+        ///
+        /// The per-thread/state lock is downgraded atomically, without allowing
+        /// any writers to take exclusive access of the lock in the meantime.
+        ///
+        /// It returns an RAII guard which will drop the write access to the
+        /// state and to the store when dropped.
+        fn downgrade(self) -> RoomEventCacheStateLockReadGuard<'a> {
+            RoomEventCacheStateLockReadGuard { state: self.state.downgrade(), store: self.store }
```

Comment:
Hey, neat that you can downgrade, didn't notice that till now.

Reply:
❤️